src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,151 @@
+            protected int _streamWindowUpdateRatio;
+            protected int _delivered;
+            protected int _streamWindowSize;
+
+            protected readonly Http2Connection _connection;
+            protected readonly Http2Stream _stream;
+
+            public Http2StreamWindowManager(Http2Connection connection, Http2Stream stream)
+            {
+                _connection = connection;
+
+                _stream = stream;
+                _streamWindowSize = connection.InitialStreamWindowSize;
+                _streamWindowUpdateRatio = _connection._pool.Settings._streamWindowUpdateRatio;
+                _stream.Trace($"StreamWindowSize: {StreamWindowSize}, StreamWindowThreshold: {StreamWindowThreshold}, streamWindowUpdateRatio: {_streamWindowUpdateRatio}");
+            }
+
+            internal int StreamWindowSize => _streamWindowSize;
+
+            internal int StreamWindowThreshold => _streamWindowSize / _streamWindowUpdateRatio;
+
+            public virtual void AdjustWindow(int bytesConsumed)
+            {
+                Debug.Assert(bytesConsumed > 0);
+                Debug.Assert(_delivered < StreamWindowThreshold);
+
+                if (!_stream.ExpectResponseData)
+                {
+                    // We are not expecting any more data (because we've either completed or aborted).
+                    // So no need to send any more WINDOW_UPDATEs.
+                    return;
+                }
+
+                _delivered += bytesConsumed;
+                if (_delivered < StreamWindowThreshold)
+                {
+                    return;
+                }
+
+                int windowUpdateSize = _delivered;
+                _delivered = 0;
+
+                Task sendWindowUpdateTask = _connection.SendWindowUpdateAsync(_stream.StreamId, windowUpdateSize);
+                _connection.LogExceptions(sendWindowUpdateTask);
+            }
+        }
+
+        private class DynamicHttp2StreamWindowManager : Http2StreamWindowManager
+        {
+            private DateTime _lastWindowUpdate = DateTime.Now;
+
+            private long _magic = 1;
+
+            public DynamicHttp2StreamWindowManager(Http2Connection connection, Http2Stream stream)
+                : base(connection, stream)
+            {
+                _magic = connection._pool.Settings._streamWindowMagicMultiplier;
+                _stream.Trace($"magic:{_magic}");
+            }
+
+            public override void AdjustWindow(int bytesConsumed)
+            {
+                _delivered += bytesConsumed;
+                if (_delivered < StreamWindowThreshold)
+                {
+                    return;
+                }
+
+                TimeSpan rtt = _connection._rttEstimator!.Rtt;
+                TimeSpan dt = DateTime.Now - _lastWindowUpdate;
+
+                int windowSizeIncrement = _delivered;
+
+                if (_magic * _delivered * rtt.Ticks > StreamWindowThreshold * dt.Ticks)
```

Comment:
So you're omitting here the **DrainRatio** which you said in **4** in msquic, thus your `_magic` is basically **12.5** in the correctly growing scenario (you said it behaves OK for `_magic = 50`. Do I get I get it right?